primitives/keystore/src/lib.rs

```diff
@@ -0,0 +1,385 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2019-2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Keystore traits
+pub mod testing;
+#[cfg(feature = "std")]
+pub mod vrf;
+
+use std::sync::Arc;
+use async_trait::async_trait;
+use futures::{executor::block_on, future::join_all};
+use sp_core::{
+	crypto::{KeyTypeId, CryptoTypePublicPair},
+	ed25519, sr25519, ecdsa,
+};
+#[cfg(feature = "std")]
+use crate::vrf::{VRFTranscriptData, VRFSignature};
+
+/// CryptoStore error
+#[derive(Debug, derive_more::Display)]
+pub enum Error {
+	/// Public key type is not supported
+	#[display(fmt="Key not supported: {:?}", _0)]
+	KeyNotSupported(KeyTypeId),
+	/// Pair not found for public key and KeyTypeId
+	#[display(fmt="Pair was not found: {}", _0)]
+	PairNotFound(String),
+	/// Validation error
+	#[display(fmt="Validation error: {}", _0)]
+	ValidationError(String),
+	/// Keystore unavailable
+	#[display(fmt="Keystore unavailable")]
+	Unavailable,
+	/// Programming errors
+	#[display(fmt="An unknown keystore error occurred: {}", _0)]
+	Other(String)
+}
```

Comment:
Curious: this works for wasm? Do we have `String` support in `wasm` now? Or am I misreading the `#[cfg(feature = "std")]` and this never actually compiles to wasm?